man/systemd.exec.xml

```diff
@@ -1963,6 +1963,29 @@ BindReadOnlyPaths=/var/lib/systemd</programlisting>
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>PrivatePIDs=</varname></term>
+
+        <listitem><para>Takes a boolean argument. Defaults to false. If enabled, sets up a new PID namespace
+        for the executed processes. Each executed process is now PID 1 - the init process - in the new namespace.
+        <filename>/proc/</filename> is mounted such that only processes in the PID namespace are visible.
+        If <varname>PrivatePIDs=</varname> is set, <varname>MountAPIVFS=yes</varname> is implied.</para>
+
+        <para>Note <varname>PrivatePIDs=</varname> is not supported with <varname>Type=forking</varname> since
+        the kernel will kill all processes in the PID namespace if the init process terminates.</para>
+
+        <para>This setting will be ignored if the kernel does not support PID namespaces.</para>
+
+        <para>If this setting is enabled for an unprivileged user service and <filename>/proc</filename> is masked
+        (e.g. <filename>/proc/kmsg</filename> is over-mounted with <constant>tmpfs</constant> like
+        <citerefentry><refentrytitle>systemd-nspawn</refentrytitle><manvolnum>1</manvolnum></citerefentry> does),
+        a new <filename>/proc</filename> will not be mounted in the new PID namespace and processes from the parent
+        PID namespace will still be visible in <filename>/proc</filename>. This is due to a kernel restriction not
+        allowing unprivileged user namespaces to mount a less restrictive instance of <filename>/proc</filename>.</para>
```

Comment:
hmm, i think we should not allow this: run a service in a pidns with the host procfs mounted it, the distinctions are just too many. 

I think we should in the long run reconsider in nspawn whether we want to overmount /proc/ like this (or at least make it an option).

I think we should treat an overmounted /proc/ the same way as pidns not available due to perms really.

Reply:
@poettering yea for user experience, I agree.

The problem I was having is we unshare the PID namespace prior to the mount namespace and mounting /proc. So we don't know whether the /proc mount will fail until we're already in the PID namespace.

I have a few possible solutions:
1. Unshare the mount namespace, set up all mounts, then only if /proc mount succeeded, unshare the PID namespace + remount /proc. IMHO, this will be challenging because after calling apply_mount_namespace(), we're already in the container root filesystem and we want to avoid mutating it (e.g. create temporary mount point, mount new proc and its child mounts there, move them over).
2. Before unsharing the PID namespace, try to detect if /proc is over-mounted. This is complicated because mounts over empty dirs are allowed (e.g. /proc/sys/fs/binfmt_misc) while other mounts over non-empty dirs/files are not (e.g. /proc/kmsg). I feel this would be brittle
3. Before unsharing the PID namespace, clone a temporary short-lived process into a new mount namespace, enable slave propagation, then try to mount /proc. If /proc mount fails, we know to not unshare the PID namespace. We have some precedent for doing this as we fork off a temporary process to set up user namespaces for private users.

I vote 3 but happy to hear other opinions/options.

Reply:
yeah, option 3 sounds good to me. kinda sucks, but it is how it is.

Reply:
@poettering fyi fun fact, we have to unshare both the mount namespace AND PID namespace in the temporary subprocess. It seems the kernel doesn't let unprivileged user namespaces mount a new proc unless they're in a new PID namespace, otherwise we get more EACCES errors

Anyways, got it to work, pushing soon, sorry for adding more reviews here :/